File: src/main/java/com/projectswg/forwarder/resources/networking/packets/SequencedPacket.java
package com.projectswg.forwarder.resources.networking.packets;

import org.jetbrains.annotations.NotNull;

import java.nio.ByteBuffer;

public interface SequencedPacket extends Comparable<SequencedPacket> {
	void setSequence(short sequence);
	short getSequence();
	
	ByteBuffer encode();
	
	default int compareTo(@NotNull SequencedPacket p) {
		return compare(this, p);
	}
	
	static int compare(@NotNull SequencedPacket a, @NotNull SequencedPacket b) {
		return compare(a.getSequence(), b.getSequence());
	}
	
	static int compare(short a, short b) {
		int aSeq = a & 0xFFFF;
		int bSeq = b & 0xFFFF;
		if (aSeq == bSeq)
			return 0;
		
		int diff = bSeq - aSeq;
		if (diff <= 0)
			diff += 0x10000;
		
		return diff < 30000 ? -1 : 1;
	}
	
}

File: src/main/kotlin/com/projectswg/forwarder/resources/networking/packets/SequencedPacket.kt
package com.projectswg.forwarder.resources.networking.packets

import com.projectswg.common.network.NetBuffer

interface SequencedPacket : Comparable<SequencedPacket> {
	
	var sequence: Short
	
	fun encode(): NetBuffer
	
	override fun compareTo(other: SequencedPacket): Int {
		return compare(this, other)
	}
	
	companion object {
		
		fun compare(a: SequencedPacket, b: SequencedPacket): Int {
			return compare(a.sequence, b.sequence)
		}
		
		fun compare(a: Short, b: Short): Int {
			val aSeq = a.toInt() and 0xFFFF
			val bSeq = b.toInt() and 0xFFFF
			if (aSeq == bSeq)
				return 0
			
			var diff = bSeq - aSeq
			if (diff <= 0)
				diff += 0x10000
			
			return if (diff < 30000) -1 else 1
		}
	}
	
}
